File: src/mesh/core/wikilinks.py
"""Wikilink resolution: ``[[Title]]`` / ``[[n-id]]`` / ``[[t-id]]`` → ``related``.

A note body may reference other entities with ``[[…]]`` links. Resolution turns
those links into the note's ``related`` id list without ever touching the body:
:func:`resolve_wikilinks` returns the body **verbatim** alongside the resolved
ids, so it is a pure derivation and running it twice is a no-op.

Two link shapes:

* **Id form** — ``[[n-id]]`` / ``[[t-id]]`` pass through directly; the id is
  taken as-is, with no file lookup (an id-only body never reads disk).
* **Title form** — ``[[Some Title]]`` is resolved by an on-disk scan of
  ``notes/`` for a mesh note whose ``title`` matches exactly (after stripping
  surrounding whitespace). Only mesh-owned notes (id ``n-…``) are indexed,
  mirroring ``list_notes`` — a coexisting foreign file (any writer sharing the
  folder) never shadows a link nor leaks a foreign id into ``related``.

Either shape may carry the dialect's decorations — a display alias after ``|``
and/or a heading (``#``) / block (``^``) anchor — which name a label for, or a
location inside, the target rather than a different entity. They are stripped by
:func:`_normalise_target` before lookup, so ``[[Note|display]]``,
``[[Note#Section]]``, ``[[Note^block]]`` and ``[[Note#Section|display]]`` all
resolve exactly as ``[[Note]]`` does, and a broken one is reported by its entity
name rather than by text still carrying a pipe or hash.

The match is an **exact** title comparison, deliberately unlike the
slug-normalized matching that ``core.notes`` uses for CLI ``<id|slug>`` args:
wikilinks name a note's title; slugs are user shorthand.

An unresolvable title is left verbatim in the body and reported by
:func:`find_dangling` for ``mesh status``. Resolution needs no daemon — it is a
direct filesystem read. A note whose body links its *own* title resolves to
itself (it is already on disk); that is allowed and harmless.
"""

from __future__ import annotations

import itertools
import re
from collections.abc import Iterator
from pathlib import Path

from mesh.storage.files import iter_md, read_post

# A ``[[…]]`` link: capture the inner text, excluding brackets and newlines.
_WIKILINK = re.compile(r"\[\[([^\[\]\n]+?)\]\]")
# Id-form link: a mesh id prefix (``n-``/``t-``) followed by id characters.
_ID_FORM = re.compile(r"^[nt]-[0-9A-Za-z]+$")
_MESH_ID_PREFIX = "n-"
_MESH_ID_PREFIXES = ("n-", "t-")
# The alias separator and the two anchor sigils of the wikilink dialect. Kept
# next to ``_WIKILINK`` because they are the same grammar: what the brackets
# capture, :func:`_normalise_target` reduces to the entity name.
_ALIAS_SEPARATOR = "|"
_ANCHOR_SIGILS = ("#", "^")
# The task lifecycle folders, walked non-recursively — the same two folders and
# the same shape as ``core.tasks._TASK_SUBDIRS`` / ``core.tasks.in_task_scope``.
# Spelled here rather than imported because ``core.tasks`` imports *this* module
# (:func:`resolve_wikilinks` runs on every task body write); the two must be kept
# in step, and ``tests/notes/test_wikilinks.py`` pins that they select the same
# files.
_TASK_SUBDIRS: tuple[str, ...] = ("open", "done")


def _notes_root(vault_path: Path) -> Path:
    return vault_path / "notes"


def _tasks_root(vault_path: Path) -> Path:
    return vault_path / "tasks"


def _iter_note_files(vault_path: Path) -> Iterator[Path]:
    """Yield every ``*.md`` under ``notes/``, sorted.

    A thin, deterministically-ordered call onto the one shared vault walk
    (:func:`mesh.storage.files.iter_md`).
    """
    yield from sorted(iter_md(_notes_root(vault_path)))


def _iter_task_files(vault_path: Path) -> Iterator[Path]:
    """Yield every ``*.md`` directly under ``tasks/open/`` then ``tasks/done/``, sorted.

    Non-recursive over exactly those two folders, mirroring
    :func:`mesh.core.tasks._iter_task_files` and
    :func:`mesh.core.tasks.in_task_scope`: those folders *are* the task
    lifecycle, so a file filed beside them (``tasks/archive/t-z.md``) or a level
    deeper (``tasks/open/sub/t-z.md``) is not a task any verb can resolve, get,
    claim or edit. Walking it anyway would let ``mesh status`` count dangling
    links from files the tool offers no way to fix — a health signal naming a
    problem outside its own scope. Sorted per folder, so the scan order is
    deterministic and ``open`` precedes ``done`` exactly as in ``core.tasks``.
    """
    root = _tasks_root(vault_path)
    for sub in _TASK_SUBDIRS:
        yield from sorted(iter_md(root / sub, recursive=False))


def _normalise_target(text: str) -> str:
    """Reduce one raw ``[[…]]`` inner text to the entity name it addresses.

    The dialect the surrounding Markdown ecosystem writes is wider than
    ``[[Title]]``: a link may carry a display alias after ``|``
    (``[[Note|display]]``) and/or a heading (``[[Note#Section]]``) or block
    (``[[Note^block-id]]``) anchor, in any combination
    (``[[Note#Section|display]]``). All three decorations name a *location
    inside* the target or a *label for* it — never a different entity — so they
    are stripped before lookup: alias first (everything after the first ``|``
    is display text, anchors included), then everything from the first ``#`` or
    ``^``. A link carrying none of them is returned unchanged, so the plain-link
    path is byte-identical to what it was before this normalisation existed.

    A bare anchor (``[[#Heading]]``, ``[[^block]]``) names a spot in the *current*
    file and reduces to ``""``; :func:`_link_targets` drops those, so they neither
    resolve into ``related`` nor count as dangling.

    The trade-off is explicit: a note whose *title* contains ``|``, ``#`` or ``^``
    can no longer be linked by title, because the dialect gives no way to escape
    them — the same restriction every host app that writes this syntax imposes.
    """
    target = text.split(_ALIAS_SEPARATOR, 1)[0]
    for sigil in _ANCHOR_SIGILS:
        target = target.split(sigil, 1)[0]
    return target.strip()


def _link_targets(body: str) -> list[str]:
    """Return the normalised target of every ``[[…]]`` link, in body order.

    The one boundary where raw link text becomes a lookup key — both
    :func:`resolve_wikilinks` and :func:`find_dangling` read through here, so the
    dialect they accept and the text ``mesh status`` reports can never drift
    apart. Empty targets (a bare ``[[#anchor]]``, or an ``[[|display]]`` with no
    name) are dropped: they address no entity.
    """
    targets = (_normalise_target(match.group(1)) for match in _WIKILINK.finditer(body))
    return [target for target in targets if target]


def _title_index(vault_path: Path) -> dict[str, str]:
    """Map exact ``title`` → mesh ``id`` for every mesh note under ``notes/``.

    Only files whose ``id`` starts with ``n-`` are indexed (foreign files — any
    writer sharing the folder — are skipped, exactly as ``list_notes`` surfaces
    only mesh notes). On a
    duplicate title the first file in sorted order wins, keeping resolution
    deterministic.
    """
    index: dict[str, str] = {}
    for path in _iter_note_files(vault_path):
        post = read_post(path)
        if post is None:
            continue
        meta = post.metadata
        note_id = meta.get("id")
        title = meta.get("title")
        if (
            isinstance(note_id, str)
            and note_id.startswith(_MESH_ID_PREFIX)
            and isinstance(title, str)
            and title not in index
        ):
            index[title] = note_id
    return index


def resolve_wikilinks(body: str, vault_path: Path) -> tuple[str, list[str]]:
    """Resolve ``[[…]]`` links in ``body`` to a ``related`` id list.

    Returns ``(body, related)`` where ``body`` is returned unchanged (links are
    never rewritten) and ``related`` holds the resolved ids in first-seen order
    with duplicates dropped. Id-form links (``[[n-id]]`` / ``[[t-id]]``) resolve
    directly; title links resolve via an on-disk lookup of mesh notes; a title
    with no match is skipped (see :func:`find_dangling`). The title index is
    built lazily, so a body of only id-form links reads no files.
    """
    index: dict[str, str] | None = None
    related: list[str] = []
    for target in _link_targets(body):
        if _ID_FORM.match(target):
            resolved = target
        else:
            if index is None:
                index = _title_index(vault_path)
            found = index.get(target)
            if found is None:
                continue  # dangling: leave verbatim, omit from related
            resolved = found
        if resolved not in related:
            related.append(resolved)
    return body, related


def find_dangling(vault_path: Path) -> list[str]:
    """Return the unresolvable ``[[Title]]`` link texts across the whole vault.

    Scans every mesh note **and task** body (``notes/**`` recursively, plus
    ``tasks/open/`` and ``tasks/done/`` non-recursively — exactly the folder scope
    ``core.tasks`` resolves through, so every file counted here is one a verb can
    actually reach) for title-form links whose target matches no mesh note,
    de-duplicated in first-seen order (sorted file scan per root, notes before
    tasks, body order within a file). Targets are reported *normalised* (see
    :func:`_normalise_target`), so a broken ``[[Ghost#Sec|display]]`` is reported
    as ``Ghost`` — the entity that is missing, not the raw link text.
    Id-form links are never dangling. The title *index* itself
    stays notes-only — titles resolve to notes by contract (see module docstring)
    — only the scan for link *targets* widens to cover tasks. Consumed by
    ``mesh status`` to surface broken references across the whole corpus (root
    tech.md § B6).
    """
    index = _title_index(vault_path)
    dangling: list[str] = []
    for path in itertools.chain(_iter_note_files(vault_path), _iter_task_files(vault_path)):
        post = read_post(path)
        if post is None:
            continue
        note_id = post.metadata.get("id")
        if not (isinstance(note_id, str) and note_id.startswith(_MESH_ID_PREFIXES)):
            continue
        for target in _link_targets(post.content):
            if _ID_FORM.match(target):
                continue
            if target not in index and target not in dangling:
                dangling.append(target)
    return dangling

File: src/domain/wikilinks.rs
//! Wikilink derivation: `[[target]]` in a body becomes `related` ids. Bodies are never rewritten.

use std::collections::HashMap;

use crate::config::Config;
use crate::fm::read_meta_only;
use crate::model::common::{meta_str, meta_strings};
use crate::spaces::Space;
use crate::storage::iter_md;
use crate::text::{is_id_form, link_targets};

/// The exact-title index over the notes space: `title -> n-id`, first sorted file wins.
///
/// Titles resolve to notes by contract; memory and asset titles are sentences and would
/// collide constantly, so they are never title targets.
pub fn title_index(cfg: &Config) -> HashMap<String, String> {
    let mut index: HashMap<String, String> = HashMap::new();
    let Ok(root) = cfg.root(Space::Notes) else {
        return index;
    };
    for path in iter_md(root, true, cfg.spaces.exclusions_for(Space::Notes)) {
        let Some(meta) = read_meta_only(&path) else {
            continue;
        };
        let (Some(id), Some(title)) = (meta_str(&meta, "id"), meta_str(&meta, "title")) else {
            continue;
        };
        if !id.starts_with("n-") {
            continue;
        }
        index
            .entry(title.to_string())
            .or_insert_with(|| id.to_string());
    }
    index
}

/// Derive `related` from a body. The body is returned to the caller untouched.
///
/// Id-form targets pass through with no file lookup; title-form targets resolve against the
/// notes index, which is built lazily and only when a title-form target is present.
pub fn resolve_wikilinks(cfg: &Config, body: &str) -> Vec<String> {
    let targets = link_targets(body);
    if targets.is_empty() {
        return Vec::new();
    }
    let mut index: Option<HashMap<String, String>> = None;
    let mut related: Vec<String> = Vec::new();
    for target in targets {
        let resolved = if is_id_form(&target) {
            target
        } else {
            let idx = index.get_or_insert_with(|| title_index(cfg));
            match idx.get(&target) {
                Some(id) => id.clone(),
                None => continue,
            }
        };
        if !related.contains(&resolved) {
            related.push(resolved);
        }
    }
    related
}

/// Every link target in the vault that resolves to nothing, first-seen order, deduped.
///
/// Scans notes (recursive) then `tasks/open` and `tasks/done` (non-recursive), id-bearing
/// files only.
pub fn find_dangling(cfg: &Config) -> Vec<String> {
    let index = title_index(cfg);
    let mut out: Vec<String> = Vec::new();
    for path in scan_paths(cfg) {
        let Some(doc) = crate::fm::read_doc(&path) else {
            continue;
        };
        let Some(id) = meta_str(&doc.meta, "id") else {
            continue;
        };
        if !(id.starts_with("n-") || id.starts_with("t-")) {
            continue;
        }
        for target in link_targets(&doc.body) {
            if is_id_form(&target) || index.contains_key(&target) || out.contains(&target) {
                continue;
            }
            out.push(target);
        }
    }
    out
}

/// The ids of notes whose body contains a title-form link to `title`.
///
/// Used by `note update --title` to warn that a rename dangles existing links.
pub fn backlinks_by_title(cfg: &Config, title: &str) -> Vec<String> {
    let mut out: Vec<String> = Vec::new();
    let Ok(root) = cfg.root(Space::Notes) else {
        return out;
    };
    for path in iter_md(root, true, cfg.spaces.exclusions_for(Space::Notes)) {
        let Some(doc) = crate::fm::read_doc(&path) else {
            continue;
        };
        let Some(id) = meta_str(&doc.meta, "id") else {
            continue;
        };
        if !id.starts_with("n-") {
            continue;
        }
        if link_targets(&doc.body).iter().any(|t| t == title) && !out.contains(&id.to_string()) {
            out.push(id.to_string());
        }
    }
    out
}

/// The ids named by another entity's `related` list, keyed by target id (the inbound index).
pub fn inbound_from_rows(rows: &[crate::fm::Row]) -> HashMap<String, Vec<String>> {
    let mut index: HashMap<String, Vec<String>> = HashMap::new();
    for row in rows {
        let Some(id) = meta_str(&row.meta, "id") else {
            continue;
        };
        for target in meta_strings(&row.meta, "related") {
            index.entry(target).or_default().push(id.to_string());
        }
    }
    for sources in index.values_mut() {
        sources.sort();
        sources.dedup();
    }
    index
}

fn scan_paths(cfg: &Config) -> Vec<std::path::PathBuf> {
    let mut paths: Vec<std::path::PathBuf> = Vec::new();
    if let Ok(notes) = cfg.root(Space::Notes) {
        paths.extend(iter_md(
            notes,
            true,
            cfg.spaces.exclusions_for(Space::Notes),
        ));
    }
    if let Ok(tasks) = cfg.root(Space::Tasks) {
        for sub in ["open", "done"] {
            paths.extend(iter_md(&tasks.join(sub), false, &[]));
        }
    }
    paths
}

#[cfg(test)]
#[allow(clippy::unwrap_used, clippy::expect_used, clippy::panic)]
mod tests {
    use super::*;
    use crate::config::test_support::config_for;
    use crate::config::Config;

    fn vault_with(files: &[(&str, &str)]) -> (tempfile::TempDir, Config) {
        let dir = tempfile::tempdir().unwrap();
        for (rel, contents) in files {
            let path = dir.path().join(rel);
            if let Some(parent) = path.parent() {
                std::fs::create_dir_all(parent).unwrap();
            }
            std::fs::write(path, contents).unwrap();
        }
        let cfg = config_for(dir.path());
        (dir, cfg)
    }

    fn note(id: &str, title: &str, body: &str) -> String {
        format!("---\nid: {id}\ntype: note\ntitle: {title}\n---\n\n{body}\n")
    }

    #[test]
    fn title_index_is_exact_and_first_sorted_file_wins() {
        let (_d, cfg) = vault_with(&[
            ("notes/n-A.md", &note("n-A", "Alpha", "")),
            ("notes/n-B.md", &note("n-B", "Alpha", "")),
            ("notes/foreign.md", "# Alpha\n"),
        ]);
        let index = title_index(&cfg);
        assert_eq!(index.get("Alpha").map(String::as_str), Some("n-A"));
        assert_eq!(index.len(), 1);
    }

    #[test]
    fn id_form_passes_through_without_a_lookup() {
        let (_d, cfg) = vault_with(&[]);
        assert_eq!(
            resolve_wikilinks(&cfg, "see [[t-99]] and [[m-1]]"),
            ["t-99", "m-1"]
        );
    }

    #[test]
    fn title_form_resolves_and_dangling_is_omitted() {
        let (_d, cfg) = vault_with(&[("notes/n-A.md", &note("n-A", "Alpha", ""))]);
        assert_eq!(
            resolve_wikilinks(&cfg, "[[Alpha|x]] and [[Ghost]] and [[Alpha#h]]"),
            ["n-A"]
        );
    }

    #[test]
    fn find_dangling_reports_normalised_targets_once() {
        let (_d, cfg) = vault_with(&[
            (
                "notes/n-A.md",
                &note("n-A", "Alpha", "[[Ghost#Sec|display]] [[Alpha]] [[Ghost]]"),
            ),
            ("notes/foreign.md", "[[Invisible]]"),
        ]);
        assert_eq!(find_dangling(&cfg), ["Ghost"]);
    }

    #[test]
    fn backlinks_by_title_finds_the_linkers() {
        let (_d, cfg) = vault_with(&[
            ("notes/n-A.md", &note("n-A", "Alpha", "")),
            ("notes/n-B.md", &note("n-B", "Beta", "link to [[Alpha]]")),
        ]);
        assert_eq!(backlinks_by_title(&cfg, "Alpha"), ["n-B"]);
        assert!(backlinks_by_title(&cfg, "Nobody").is_empty());
    }
}
